File: python/src/aoc/year2024/day12.py
from aoc.util import load_input, load_example


def read_grid(lines):
    grid = {}
    for y, line in enumerate(lines):
        for x, value in enumerate(line.strip()):
            grid[x, y] = value
    return grid


def part1(lines):
    """
    >>> part1(load_example(__file__, "5"))
    1930
    """
    grid = read_grid(lines)
    total = 0
    closed_set = set()
    for start in grid:
        if start in closed_set:
            continue
        plant = grid[start]
        open_heap = [start]
        area = 1
        sx, sy = start
        fences = 4 - sum(
            (m in grid and grid[m] == plant) for m in ((sx - 1, sy), (sx + 1, sy), (sx, sy - 1), (sx, sy + 1))
        )
        while open_heap:
            position = open_heap.pop()
            if position in closed_set:
                continue
            closed_set.add(position)
            px, py = position
            for n in ((px - 1, py), (px + 1, py), (px, py - 1), (px, py + 1)):
                if n not in grid:
                    continue
                if n in closed_set:
                    continue
                if grid[n] == plant and n not in open_heap:
                    area += 1
                    nx, ny = n
                    fences += 4 - sum(
                        (m in grid and grid[m] == plant)
                        for m in ((nx - 1, ny), (nx + 1, ny), (nx, ny - 1), (nx, ny + 1))
                    )
                    open_heap.append(n)
        total += area * fences

    return total


def merge_fences(fences):
    new_fences = []
    while fences:
        d, (px, py), (nx, ny) = fences.pop()
        if d == "W" or d == "E":
            i = 1
            while True:
                c = (d, (px, py + i), (nx, ny + i))
                if c in fences:
                    fences.remove(c)
                else:
                    break
                i += 1
            i = -1
            while True:
                c = (d, (px, py + i), (nx, ny + i))
                if c in fences:
                    fences.remove(c)
                else:
                    break
                i -= 1
        if d == "N" or d == "S":
            i = 1
            while True:
                c = (d, (px + i, py), (nx + i, ny))
                if c in fences:
                    fences.remove(c)
                else:
                    break
                i += 1
            i = -1
            while True:
                c = (d, (px + i, py), (nx + i, ny))
                if c in fences:
                    fences.remove(c)
                else:
                    break
                i -= 1
        new_fences.append((d, (px, py), (nx, ny)))
    return new_fences


def part2(lines):
    """
    >>> part2(load_example(__file__, "5"))
    1206
    """
    grid = read_grid(lines)

    total = 0
    closed_set = set()
    for start in grid:
        if start in closed_set:
            continue
        plant = grid[start]
        open_heap = [start]
        area = 1
        sx, sy = start
        fences = [
            (d, start, m)
            for d, m in (("W", (sx - 1, sy)), ("E", (sx + 1, sy)), ("N", (sx, sy - 1)), ("S", (sx, sy + 1)))
            if m not in grid or grid[m] != plant
        ]
        while open_heap:
            position = open_heap.pop()
            if position in closed_set:
                continue
            closed_set.add(position)
            px, py = position
            for n in ((px - 1, py), (px + 1, py), (px, py - 1), (px, py + 1)):
                if n not in grid:
                    continue
                if n in closed_set:
                    continue
                if grid[n] == plant and n not in open_heap:
                    area += 1
                    nx, ny = n
                    fences += [
                        (d, n, m)
                        for d, m in (("W", (nx - 1, ny)), ("E", (nx + 1, ny)), ("N", (nx, ny - 1)), ("S", (nx, ny + 1)))
                        if m not in grid or grid[m] != plant
                    ]
                    open_heap.append(n)
        total += area * len(merge_fences(fences))

    return total


if __name__ == "__main__":
    data = load_input(__file__, 2024, "12")
    print(part1(data))
    print(part2(data))

File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2024/Day12.kt
package io.github.ocirne.aoc.year2024

import io.github.ocirne.aoc.AocChallenge
import io.github.ocirne.aoc.NSWE

class Day12(val lines: List<String>) : AocChallenge(2024, 12) {

    data class Position(val x: Int, val y: Int)

    class Grid<T>(val lines: List<String>, val interpretValue: (Char) -> T) {

        val width = lines.firstOrNull()?.length
        val height = lines.size
        val grid = readGrid()

        private fun readGrid(): Map<Position, T> {
            val grid = mutableMapOf<Position, T>()
            lines.mapIndexed { y, line ->
                line.mapIndexed { x, value ->
                    grid[Position(x, y)] = interpretValue(value)
                }
            }
            return grid
        }

        fun neighbors4S(p: Position): Sequence<Position> {
            val (x, y) = p
            return sequence {
                for ((dx, dy) in NSWE) {
                    if (x + dx in 0 until width!! && y + dy in 0 until height) {
                        yield(Position(x + dx, y + dy))
                    }
                }
            }
        }

        operator fun get(p: Position): T {
            return grid[p]!!
        }
    }

    val grid = Grid(lines) { it }

    override fun part1(): Long {
        val plants = grid.grid.values.toSet()
        val areas = grid.grid.map { (_, plant) -> plant }.groupingBy { it }.eachCount()
        val fences = mutableMapOf<Char, Int>()
        grid.grid.map { (position, plant) ->
            val nonNeigbor = 4 - grid.neighbors4S(position).filter { neighbor -> grid[neighbor] == plant }.count()
            fences[plant] = fences.getOrDefault(plant, 0) + nonNeigbor
        }
        println(plants)
        println(areas)
        println(fences)
        return -1
    }

    override fun part2(): Long {
        return -1
    }
}
